### ISSUE TITLE
Handling CTRL+C (SIGINT) in an executor does not work anymore after upgrading NX

### ISSUE THREAD

**Issue by pct-cclausen** [nrwl/nx#23585]
### Current Behavior

I have an executor in our nx monorepo which is responsible for managing "docker compose up" for a project, then running cypress tests against the docker containers and then shutdown everything.

To handle a developer pressing CTRL+C during a running test I use the "death" npm package, which installs hooks that allow the program to handle SIGINT/SIGTERM/etc and postpone program exit until these are completed.

This exists in 2 monorepos in fact: 
- Our own, which is at NX@17.1.3. Everything works fine here.
- A repo of one of our customers. This until recently used NX 16.3.2

I am upgrading packages in the customers repository. Specifically I have migrated to NX 18.2.4, as higher versions are not supported by some 3rd party packages we use.

Since this upgrade the shutdown handling does not work correctly anymore. It seems my executor code is shutdown before it can finish or even start `docker compose down`, which causes docker containers to stay online, which is bad for developer experience.

### Expected Behavior

It should not stop the executor process prematurely.
The example repo should print "WORK DONE" after ~3 seconds.

### GitHub Repo

https://github.com/pct-digital/typescript-homework/tree/on_death_executor_issue

### Steps to Reproduce

See the README in the repo, branch on_death_executor_issue.

### Nx Report

```shell
NX   Report complete - copy this into the issue template

Node   : 20.13.1
OS     : linux-x64
npm    : 9.8.1

nx                 : 19.0.2
@nx/js             : 19.0.2
@nx/jest           : 19.0.2
@nx/linter         : 19.0.2
@nx/eslint         : 19.0.2
@nx/workspace      : 19.0.2
@nx/devkit         : 19.0.2
@nx/eslint-plugin  : 19.0.2
@nx/node           : 19.0.4
@nx/plugin         : 19.0.2
@nrwl/tao          : 19.0.2
@nx/web            : 19.0.2
@nx/webpack        : 19.0.2
typescript         : 5.4.5
---------------------------------------
Registered Plugins:
@nx/webpack/plugin
@nx/eslint/plugin
@nx/jest/plugin
---------------------------------------
Local workspace plugins:
         @nx-ts/pct-pipeline
---------------------------------------
The following packages should match the installed version of nx
  - @nx/node@19.0.4
  - @nrwl/node@19.0.4

To fix this, run `nx migrate nx@19.0.4`
```

### Operating System

- [ ] macOS
- [X] Linux
- [ ] Windows
- [ ] Other (Please specify)

### Additional Information

I was able to "fix" the issue by manually rolling back the version of the NX package to 17.2.8 and running npm install.
I am however not very happy with letting it stay like that, I think the versions should all match?

But this might be helpful for debugging the issue:
- Setting NX to 17.2.8 makes it work again in the customers repository.
- 17.3.0 is the first version that fails
- My example monorepo is based on some minimal repo setup I had lying around, it already uses NX 19, still does not work. Setting the NX version to 17.2.8 here causes lots of errors, I guess that version is too old to work with NX 19.
- I tried to setup an NX repo at version 17.2.8 directly, but that somehow also does not work.

Maybe it has something to do with changes to this file: https://github.com/nrwl/nx/blame/17.3.0/packages/nx/src/tasks-runner/forked-process-task-runner.ts between 17.2.8 and 17.3.0

The only working examples of my code exist in our own private nx monorepo which I cannot share, somehow I have not been able to recreate a working minimal example, only a non-working one (see above).

**Comment by pct-cclausen** [comment]
I've now created a workaround that no longer depends on the complex interplay of signals and the death package.
Instead I create a new process "watch for the process with pid X, if it is no longer there run command Y".

X is the pid of the running executor
Y is "kill -- -Z" where Z is the pid of some external process like Cypress
Y also is "docker down"

Works great for my case.

Just FYI if somebody with a similar pain finds this.

This is js file for that process:
```javascript
// A helper program meant to clean up resources by running arbitrary commands
// as soon as a specific other program is not longer active

// This script is used in the acceptance-test-executor

// call like this: nde reaper.js P X1 X2 X3 X4 ...
// where P is a pid that is watched
// as soon as P can no longer be found the comamnds X1, ... XN are executed
// then reaper.js shuts down

// additionally a timeout of some hours is used, just in case.

const { execSync } = require('child_process');

const TIMEOUT = 6 * 3600 * 1000;

const targetPid = Number(process.argv[2]);
const cleanupCommands = process.argv.slice(3);

if (Number.isNaN(targetPid)) {
  throw new Error('target pid is not a number');
}

console.log('Watching process ' + targetPid + ', waiting to cleanup resources once it is gone');

function cleanup() {
  for (const cmd of cleanupCommands) {
    try {
      console.log('Cleanup... ' + cmd);
      execSync(cmd);
      console.log('...done');
    } catch (e) {
      console.log("Failed to run cleanup command '" + cmd + "'", e);
    }
  }
}

function isPidActive(pid) {
  try {
    /**
     * "As a special case, a signal of 0 can be used to test for the existence of a process."
     * So "kill" does not kill here
     */
    process.kill(targetPid, 0);
    return true;
  } catch (e) {
    return false;
  }
}

setTimeout(cleanup, TIMEOUT);

(async () => {
  for (;;) {
    await new Promise((resolve) => setTimeout(resolve, 500));

    const isActive = isPidActive(targetPid);

    if (!isActive) {
      console.log('Process ' + targetPid + ' is not active anymore. Cleaning up resources!');
      cleanup();
      process.exit(0);
    }
  }
})();
```

**Comment by dalton-oliveira** [comment]
I'm experiencing this same problem when running `nx watch --project=... -- yarn nx run ...`.  Hitting `Ctrl + c` doesn't kill the process on WSL 2, nx v20.1.0.

**Comment by dave-kooply** [comment]
NX version latest (20.3.3)

I experience similar case on me local environment with my Node js service, when I run the "serve" task with watch option on,
all work fine until I ctrl+c (SIGINT).
The node process is killed before reaching my termination code (expected!), in the terminal you see the process exit BUT in reality the process stay alive on keep cpu 100% until i Force exit it.
The issue is even worse, if I don't force exit the process on each "serve" another process get stuck.

The solution is to check if the termination signal is (SIGINT) i just process.exit it (can become an issue when you debug production on local env).

On 'htop' I see the following cmd of the process:
/Users/d/.nvm/versions/node/v20.18.1/bin/node -r /Users/d/k/b/node_modules/@nx/js/node_modules/source-map-support/register.js --inspect=localhost:9229 /Users/d/k/b/node_modules/@nx/js/src/executors/node/node-with-require-overrides

On 'Activity Monitor' i see the following cmd of the process:
/Users/d/k/b
/Users/d/.nvm/versions/node/v20.18.1/bin/node
/Library/Preferences/Logging/.plist-cache.P4y2oyOq
/Users/d/k/b/node_modules/@msgpackr-extract/msgpackr-extract-darwin-arm64/node.napi.glibc.node
/Users/d/k/b/node_modules/sharp/build/Release/sharp-darwin-arm64v8.node
/Users/d/k/b/node_modules/sharp/vendor/8.14.5/darwin-arm64v8/lib/libvips-cpp.42.dylib

**Comment by jackslooney** [comment]
Also experiencing something similar. NX version 17.2.8, so it is interesting that rolling back to this version fixed some issues for you.

I have a frontend project (web) and backend project (api) running in parallel with 
`yarn nx run-many --target=serve --projects=web,api`

When sending ctrl+c (SIGINT), the frontend node processes kill, however the backend processes do not. I can manually kill the backend process with 
`kill <process-number>`
however, this leaves orphaned related processes that stay running, until I rebuild my dev container.

This also creates a problem while editing the backend while the backend processes are running, as the hot-reloader can't kill the old processes. The result is a duplicate process that starts up, but cannot connect to my test database, as the old process is hogging the assigned port.

**Comment by forivall** [comment]
I think i'm having the same issue, except i'm primarily using it for <kbd>ctrl+z</kbd>. My current workaround is to use `kill -STOP $(pgrep -f 'nx run project:target:configuration')` from a different terminal. Then zsh reports `[1]  + 26110 suspended (signal)` and returns interactive control, where i can call `reset` in the terminal, and then after `fg`, <kbd>ctrl+z</kbd> works fine. (I suspend `--watch` tasks so i dont get a big heavy rebuild that slows down my computer when i'm working in code, and so i get a faster resume when i `fg` it and i want a rebuild)

I think it's clear that something in the output is writing a terminal control sequence which makes the terminal treat <kbd>ctrl+c</kbd> and other such commands as raw input. This happens with all tasks, so the problem is likely something in nx.

**Comment by garoyeri** [comment]
I'm on NX 21.1 and I'm seeing this issue with nx-dotnet running Aspire. When I CTRL+C, I expect Aspire to get the signal and clean up all the containers it started. However, when I do CTRL+C, it kills the process and none of the cleanup happens. It sounds like this is a similar issue. If I run the Aspire dotnet host directly, it cleans up everything correctly, meaning it received the signal.

**Comment by garoyeri** [comment]
I see this still open issue on NPM: https://github.com/npm/cli/issues/7511 which seems to indicate that it is a bug in NPM that is manifesting here too.

**Comment by leosvelperez** [comment]
This was fixed in Nx 23.0.0 via #33655, which reworked task shutdown to use a native graceful process tree kill.

On Ctrl+C / SIGINT, Nx now signals tasks leaf-first and waits for their cleanup handlers to run (e.g. `docker compose down`), and protects cleanup subprocesses spawned during shutdown from being force-killed. Earlier versions killed the process tree flat, so cleanup was cut off before it finished. The grace period is configurable via `NX_PROCESS_KILL_GRACE_PERIOD` (default 5000ms).

Please upgrade to Nx 23.0.0+ to get the fix. If you still have an issue after that, please create a new issue with a repro and as much information as possible, since this kind of issue is hard to troubleshoot.
